aws: Add an experimental IPv6 dev/test environment

russellb opened the pull request:
This PR includes support for an IPv6 dev/test environment on AWS.  It is off by default and enabled only by setting the `OPENSHIFT_INSTALL_AWS_USE_IPV6` environment variable.  See `docs/dev/aws_ipv6.md` for some more information.

The intention is to not change any behavior by default and just have this as something developers can turn on as we work across OpenShift to resolve IPv6 issues.

abhinavdahiya commented:
/hold

this needs an enhancement, and also no TF_VARs based control should be used.

russellb commented:
Fixed an issue pointed out by CI, dropped the usage of a TF_VAR_ env var, and added an enhancement here: https://github.com/openshift/enhancements/pull/87

lihongan commented on docs/dev/aws_ipv6.md:
```diff
@@ -0,0 +1,61 @@
+  IPv6, other than classic load balancers with EC2-Classic, and not EC2-VPC.
+  AWS Application Load Balancers supposedly support IPv6, but that would
+  require doing HTTPS load balancing for the API instead of just TCP load
+  balancing, so we just use the IPv4 NLBs.  API access within the cluster is
```
Looks conflict with https://github.com/openshift/enhancements/pull/87/files#diff-7dbc8a4e63bd5b199e8c113fbc62609cR72, will use ELB for API or not?

russellb replied:
Ah yeah, the enhancement is out of date.  This doc is current.

abhinavdahiya commented on data/data/aws/route53/base.tf:
```diff
@@ -85,11 +85,28 @@ resource "aws_route53_record" "etcd_a_nodes" {
+  # If the expression in the following list itself returns a list, remove the
+  # brackets to avoid interpretation as a list of lists. If the expression
+  # returns a single list item then leave it as-is and remove this TODO comment.
+  records = [var.etcd_ipv6_addresses[count.index]]
```
can we fix the TODO?

russellb replied:
See the A record above this addition, this was copied.

abhinavdahiya replied:
can we fix both then?

russellb replied:
Sure.  The code is right as is.  It looks like the comments just need to be removed.

abhinavdahiya commented on data/data/aws/route53/base.tf:
```diff
@@ -85,11 +85,28 @@ resource "aws_route53_record" "etcd_a_nodes" {
   zone_id = aws_route53_zone.int.zone_id
   name    = "_etcd-server-ssl._tcp"
-  records = formatlist("0 10 2380 %s", aws_route53_record.etcd_a_nodes.*.fqdn)
+  records = var.use_ipv6 == false ? formatlist("0 10 2380 %s", aws_route53_record.etcd_a_nodes.*.fqdn) : formatlist("0 10 2380 %s", aws_route53_record.etcd_aaaa_nodes.*.fqdn)
```
maybe we can collect the fqdn instead of switching on entire format list?

russellb replied:
I don't think I understand the suggestion here.  Can you expand on it a bit?

abhinavdahiya replied:
you are creating a formated lists in both conditionals, maybe you can condition the input list to formatlist?

russellb replied:
Oh, I see.  Sure, I'll try that.

abhinavdahiya commented on data/data/aws/route53/variables.tf:
```diff
@@ -14,6 +14,12 @@ variable "etcd_ip_addresses" {
+variable "etcd_ipv6_addresses" {
+  description = "List of string IPs (IPv6) for machines running etcd members."
+  type        = list(string)
+  default     = []
```
is the default required? if yes can you document what happens when empty vs non-zero?

russellb replied:
This is exactly like `etcd_ip_addresses` above

abhinavdahiya replied:
also similar question to https://github.com/openshift/installer/pull/2555#discussion_r354966038

russellb replied:
Sure, I can combine this one.

abhinavdahiya commented on data/data/aws/vpc/outputs.tf:
```diff
@@ -6,6 +6,10 @@ output "vpc_cidrs" {
   value = [data.aws_vpc.cluster_vpc.cidr_block]
 }
 
+output "vpc_ipv6_cidrs" {
```
why not use `vpc_cidrs` ?

russellb replied:
Because the VPC has both IPv4 and IPv6 CIDRs

abhinavdahiya replied:
my question was does the output need to make the distinct? can they be a single list?

russellb replied:
It wouldn't be easy to just use a single list.  They are used as input to different options on a resource: `cidr_blocks` vs `ipv6_cidr_blocks`, so we have to explicitly be using only the IPv4 addresses or only the IPv6 addresses.

abhinavdahiya commented on data/data/aws/vpc/sg-master.tf:
```diff
@@ -253,3 +313,33 @@ resource "aws_security_group_rule" "master_ingress_services_udp" {
 
+# For our AWS IPv6 environment, we run CoreDNS with host networking,
+# because it must use IPv4 to reach AWS DNS, so it can't be on our
+# IPv6 only SDN.
```
TODO(@abhinavdahiya): get clarity on this

russellb replied:
Are you asking me for clarity?

This issue affects several pods.  They try to reach the IPv4-only AWS APIs and can't do it from the IPv6-only SDN.  We're also looking adding a secondary interface with multus that provides IPv4 connectivity.  For now, we have a couple of pods using host networking.

abhinavdahiya replied:
https://github.com/openshift/enhancements/pull/87#issuecomment-562665545

russellb replied:
I added more details about this in the enhancement and the docs in this PR.

abhinavdahiya commented on data/data/aws/vpc/vpc-private.tf:
```diff
@@ -33,6 +48,9 @@ resource "aws_subnet" "private_subnet" {
 
   availability_zone = var.availability_zones[count.index]
 
+  ipv6_cidr_block                 = var.use_ipv6 == true ? cidrsubnet(data.aws_vpc.cluster_vpc.ipv6_cidr_block, 8, count.index) : ""
```
https://github.com/openshift/installer/pull/2555/files#diff-2f4127d65f2b75935b1181f433f7bb75R47 looks different from that one?

russellb replied:
I can't tell what that link was to.  Which other change are you referring to?

abhinavdahiya replied:
the IPV4 cidr is calculated as also similar question to https://github.com/openshift/installer/pull/2555#discussion_r354966038

my preference would be to keep it similar..

russellb replied:
I've added comments to hopefully help explain what this is doing.  Let me know if you have a more specific suggestion, since I'm not sure what you mean by wanting it similar to IPV4.

abhinavdahiya commented on data/data/aws/vpc/vpc-private.tf:
```diff
@@ -24,6 +24,21 @@ resource "aws_route" "to_nat_gw" {
 }
 
+# We can't target the NAT gw for our "private" IPv6 subnet.  Instead, we target the internet gateway,
+# since we want our private IPv6 addresses to be able to talk out to the internet, too.
```
TODO(@abhinavdahiya): need to get clarity on this.

russellb replied:
What kind of clarity are you looking for?  The AWS NAT gateway doesn't support IPv6.  They hand out globally routable IPv6 addresses and expect those to be used.

abhinavdahiya replied:
https://github.com/openshift/enhancements/pull/87#discussion_r354945664

russellb replied:
I've looked into doing what we discussed on the enhancement (switching to an egress only internet gateway).  A major downside is that tags are not supported.  A second potential downside is that we may end up wanting some direct connectivity to Nodes to exercise more of IPv6 with ingress, since AWS load balancers can only do IPv4.  At this point I'd suggest we leave this as-is ...

https://docs.aws.amazon.com/cli/latest/reference/ec2/create-egress-only-internet-gateway.html

https://docs.aws.amazon.com/cli/latest/reference/ec2/create-egress-only-internet-gateway.html

russellb commented:
rebased again on master.

abhinavdahiya commented:
can we make sure we have validations for:

1) ensure the ipv6 addresses are allowed only  when the feature-env is set.
2) ensure that AWS is the platform when the feature-env is set.

abhinavdahiya commented on data/data/aws/master/main.tf:
```diff
@@ -89,7 +91,16 @@ resource "aws_network_interface" "master" {
+
+# NOTE(russellb) For some reason, I was not able to access get IPv6 addresses
+# on the resource, but was able to get them using the network interface data
+# source.
```
did we open an issue with upstream?

russellb replied:
No, I had moved on once I had an easy enough workaround.  I can come back around to filing an upstream issue with a reproducer later, though.

abhinavdahiya commented on data/data/aws/master/main.tf:
```diff
@@ -4,6 +4,8 @@ locals {
   // and therefore are force to implicitly assume that the list is of aws_lb_target_group_arns_length - 1, in case there is no api_external
   target_group_arns_length = var.publish_strategy == "External" ? var.target_group_arns_length : var.target_group_arns_length - 1
+
+  public_endpoints = var.publish_strategy == "External" ? true : false
```
this doesn't seem to be used?

russellb replied:
Thanks.  This must of been related to something that has since been removed.

abhinavdahiya commented on data/data/aws/vpc/sg-master.tf:
```diff
@@ -23,6 +23,18 @@ resource "aws_security_group_rule" "master_mcs" {
+  from_port        = 22623
+  to_port          = 22623
+
+  count = var.use_ipv6 == true ? 1 : 0
```
would like to keep the resources in the following format..

```
resource xxx "yyy" {
  count  = 2

  ...
}
```

abhinavdahiya commented on data/data/aws/master/outputs.tf:
```diff
@@ -2,3 +2,6 @@ output "ip_addresses" {
   value = aws_network_interface.master.*.private_ips
 }
 
+output "ipv6_addresses" {
```
also similar question to https://github.com/openshift/installer/pull/2555#discussion_r354966038

russellb replied:
This one seems like a toss-up.  The master module does not take a `use_ipv6` input, which would be needed to figure out what to put in the `ip_addresses` output variable.  I could drop `ipv6_addresses`, but would add an input variable instead, so it doesn't seem much better.  I'll change it if you prefer it that way, though.

abhinavdahiya commented on data/data/aws/vpc/vpc-public.tf:
```diff
@@ -43,13 +43,28 @@ resource "aws_route" "igw_route" {
   cidr_block        = cidrsubnet(local.new_public_cidr_range, 3, count.index)
   availability_zone = var.availability_zones[count.index]
 
+  ipv6_cidr_block                 = var.use_ipv6 == true ? cidrsubnet(data.aws_vpc.cluster_vpc.ipv6_cidr_block, 8, count.index + length(var.availability_zones)) : ""
```
https://github.com/openshift/installer/blob/master/data/data/aws/vpc/vpc-public.tf#L50 is how the IPv4 address is calculated.

my preference would be to keep this calculated similar looking if possible.

russellb replied:
I've added some comments to hopefully help explain why this is the way it is.  Let me know if you have a more specific suggestion.

russellb commented:
btw ... I'll squash all of this once we're happy with it.  I'm still working through addressing comments and will post a clear comment indicating when it's ready for another look.  Thanks!

russellb commented:
> can we make sure we have validations for:
> 
>     1. ensure the ipv6 addresses are allowed only  when the feature-env is set.
> 
>     2. ensure that AWS is the platform when the feature-env is set.
> 
>     3. based on [openshift/enhancements#87 (comment)](https://github.com/openshift/enhancements/pull/87#discussion_r351302708), we ensure OVN is set.
> 
>     4. when the user has provided it's subnets, and the feature-env is set.. can we make sure that the subnets and VPCs have IPv6 CIDR attached

I've added these validations, and I believe either responded to or addressed all other comments.

@abhinavdahiya This is ready for another look.  Once you're happy with it, I can squash all of these commits.

russellb commented:
/retest

russellb commented:
/retest

russellb commented:
/test e2e-aws

russellb commented:
/close
